program/src/processor.rs

```diff
@@ -5313,6 +5318,286 @@ impl Processor {
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        num_open_orders: u8,
+        num_tokens_used: u8,
+        cpi_data: Vec<u8>,
+    ) -> MangoResult {
+        let num_open_orders = num_open_orders as usize;
+        let num_tokens_used = num_tokens_used as usize;
+        check!(num_tokens_used > 0 && num_tokens_used <= MAX_TOKENS, MangoErrorCode::InvalidParam)?;
+        const NUM_FIXED: usize = 5;
+        let (fixed_ais, var_ais) = array_refs![accounts, NUM_FIXED; ..;];
+        let (packed_open_orders_ais, var_ais) = var_ais.split_at(num_open_orders);
+        let (bank_ais, cpi_ais) = var_ais.split_at(num_tokens_used * 3);
+        let (cpi_prog_ai, cpi_ais) = cpi_ais.split_first().unwrap();
+
+        // First cpi account must be the program and it must not be mango program
+        check!(cpi_prog_ai.key != program_id, MangoErrorCode::InvalidParam)?;
+
+        let [
+            mango_group_ai,         // read
+            mango_account_ai,       // write
+            owner_ai,               // read & signer
+            mango_cache_ai,         // read
+            signer_ai,              // read
+        ] = fixed_ais;
+
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        check!(signer_ai.key == &mango_group.signer_key, MangoErrorCode::InvalidSignerKey)?;
+
+        let mut mango_account =
+            MangoAccount::load_mut_checked(mango_account_ai, program_id, mango_group_ai.key)?;
+        check!(owner_ai.is_signer, MangoErrorCode::SignerNecessary)?;
+        check!(
+            &mango_account.owner == owner_ai.key || &mango_account.delegate == owner_ai.key,
+            MangoErrorCode::InvalidOwner
+        )?;
+        check!(!mango_account.is_bankrupt, MangoErrorCode::Bankrupt)?;
+
+        // Take out the root bank, node bank and vault accounts and validate
+        let mut bank_sets = vec![];
+        let mut extra_user_assets = vec![];
+        let mut pre_amounts = vec![];
+        for bank_set_ais in bank_ais.chunks_exact(3) {
+            let bank_set_ais = array_ref![bank_set_ais, 0, 3];
+            let [root_bank_ai, node_bank_ai, vault_ai] = bank_set_ais;
+            let token_index = mango_group
+                .find_root_bank_index(root_bank_ai.key)
+                .ok_or(throw_err!(MangoErrorCode::InvalidRootBank))?;
+
+            let root_bank = RootBank::load_checked(root_bank_ai, program_id)?;
+
+            check!(
+                root_bank.node_banks.contains(node_bank_ai.key),
+                MangoErrorCode::InvalidNodeBank
+            )?;
+            let node_bank = NodeBank::load_mut_checked(node_bank_ai, program_id)?;
+            check_eq!(&node_bank.vault, vault_ai.key, MangoErrorCode::InvalidVault)?;
+            // pre_amounts.push(Account::unpack(&vault_ai.try_borrow_data()?)?.amount);
+            let token_account = TokenAccount::load_checked(vault_ai)?;
+            token_account.check_mango_reqs(&mango_group)?;
```

Comment:
I like these extra checks. Maybe even have a single function `TokenAccount::load_mango_vault_checked(...)` so we can't forget to call `check_mango_reqs()`? 

Should all the other uses of `Account::unpack()` be replaced? What makes this instance special?

Reply:
I just added `TokenAccount` in this branch. All other instances of `Account::unpack()` should be replaced once we test this out and make sure it's equivalent to `TokenAccount::load_checked()`. I did not wrap the checks together because I wanted TokenAccount::load_checked to be more general, but I can't think of a case where doing those extra checks would be harmful actually.